File: algorithms/dynamic-programming/partition-problem/python/can_partition.py
def can_partition(arr: list[int]) -> int:
    total = sum(arr)
    if total % 2 != 0:
        return 0
    target = total // 2
    dp = [False] * (target + 1)
    dp[0] = True
    for num in arr:
        for j in range(target, num - 1, -1):
            dp[j] = dp[j] or dp[j - num]
    return 1 if dp[target] else 0

File: algorithms/dynamic-programming/partition-problem/java/CanPartition.java
public class CanPartition {

    public static int canPartition(int[] arr) {
        int total = 0;
        for (int x : arr) total += x;
        if (total % 2 != 0) return 0;
        int target = total / 2;
        boolean[] dp = new boolean[target + 1];
        dp[0] = true;
        for (int num : arr) {
            for (int j = target; j >= num; j--) {
                dp[j] = dp[j] || dp[j - num];
            }
        }
        return dp[target] ? 1 : 0;
    }
}
